### PR TITLE
Add phrase confirmation to domain block dialog

### PR DESCRIPTION
**Pull request by noellabo**
image

Add phrase input to the confirmation dialog to inhibit accidental execution of dangerous operations.

### CODE REVIEW

**Comment by SouthFox-D** [comment]
Not everyone understands what "domain name" means (especially in other languages),  so it would be better to use "Please type {domain name} to confirm"?

**Comment by noellabo** [comment]
image

- I changed the wording. Thanks! @SouthFox-D
- Added a missing support for the part that calls the domain block
- adjusted the style for long labels

**Comment by shleeable** [comment]
This is excellent.

**Comment by noellabo** [comment]
image

Add setting to disable input requests. 

This pull request was created for people who accidentally lost a large number of following and followers. Therefore, the default is enabled.

**Comment by ClearlyClaire** [comment]
Also, I can't believe I'm only thinking about it right now, but there are multiple cases in which this may make harassment situations worse: long domain names, domain names with slurs, and international domain names in scripts you can't read or don't know how to type (sure, you can always copy-paste, but that may be prove troublesome, especially in a stressful situation and/or on mobile)

**Comment by noellabo** [comment]
The purpose is to prevent accidental execution, so let's simply ask to type `block`.

**Comment by noellabo** [comment]
In the reports of misoperation accidents so far, there is a case where the smartphone is repeatedly tapped in the pocket. In this case, simply adding weights or asking for multiple taps will not help. That's why I think it makes sense to ask for input.

I think making the button red is a good idea.

**Comment by ClearlyClaire** [comment]
I think “Please type [context-specific phrase] to confirm” with a short translatable context-specific phrase like in your screenshot would work, yeah. I'd add quotes around it. “block” or “block domain” sound good, for groups it would probably be “delete group”.

**Comment by ClearlyClaire** [comment]
I will do a proper review later, but this looks good to me!

File: app/javascript/mastodon/containers/domain_container.js [hunk 1]
```diff
@@ -21,6 +22,8 @@ const mapDispatchToProps = (dispatch, { intl }) => ({
       message: <FormattedMessage id='confirmations.domain_block.message' defaultMessage='Are you really, really sure you want to block the entire {domain}? In most cases a few targeted blocks or mutes are sufficient and preferable.' values={{ domain: <strong>{domain}</strong> }} />,
       confirm: intl.formatMessage(messages.blockDomainConfirm),
       onConfirm: () => dispatch(blockDomain(domain)),
+      destructive: true,
+      ...confirmDomainBlock && { passphrase: 'block' },
```
Comment (ClearlyClaire) [inline app/javascript/mastodon/containers/domain_container.js:26]: Maybe this should be translatable?
Also I don't find the `...expression` form very readable with a condition.

File: app/javascript/mastodon/containers/status_container.js [hunk 2]
```diff
@@ -226,6 +226,8 @@ const mapDispatchToProps = (dispatch, { intl, contextType }) => ({
       message: <FormattedMessage id='confirmations.domain_block.message' defaultMessage='Are you really, really sure you want to block the entire {domain}? In most cases a few targeted blocks or mutes are sufficient and preferable. You will not see content from that domain in any public timelines or your notifications. Your followers from that domain will be removed.' values={{ domain: <strong>{domain}</strong> }} />,
       confirm: intl.formatMessage(messages.blockDomainConfirm),
       onConfirm: () => dispatch(blockDomain(domain)),
+      destructive: true,
+      ...confirmDomainBlock && { passphrase: 'block' },
```
Comment (ClearlyClaire) [inline app/javascript/mastodon/containers/status_container.js:230]: Same wrt. `block` being translatable and `...expression` form.

File: app/javascript/mastodon/features/account_timeline/containers/header_container.js [hunk 3]
```diff
@@ -114,6 +114,8 @@ const mapDispatchToProps = (dispatch, { intl }) => ({
       message: <FormattedMessage id='confirmations.domain_block.message' defaultMessage='Are you really, really sure you want to block the entire {domain}? In most cases a few targeted blocks or mutes are sufficient and preferable. You will not see content from that domain in any public timelines or your notifications. Your followers from that domain will be removed.' values={{ domain: <strong>{domain}</strong> }} />,
       confirm: intl.formatMessage(messages.blockDomainConfirm),
       onConfirm: () => dispatch(blockDomain(domain)),
+      destructive: true,
+      ...confirmDomainBlock && { passphrase: 'block' },
```
Comment (ClearlyClaire) [inline app/javascript/mastodon/features/account_timeline/containers/header_container.js:118]: Same.

File: app/javascript/mastodon/features/status/index.js [hunk 4]
```diff
@@ -396,6 +396,8 @@ class Status extends ImmutablePureComponent {
       message: <FormattedMessage id='confirmations.domain_block.message' defaultMessage='Are you really, really sure you want to block the entire {domain}? In most cases a few targeted blocks or mutes are sufficient and preferable. You will not see content from that domain in any public timelines or your notifications. Your followers from that domain will be removed.' values={{ domain: <strong>{domain}</strong> }} />,
       confirm: this.props.intl.formatMessage(messages.blockDomainConfirm),
       onConfirm: () => this.props.dispatch(blockDomain(domain)),
+      destructive: true,
+      ...confirmDomainBlock && { passphrase: 'block' },
```
Comment (ClearlyClaire) [inline app/javascript/mastodon/features/status/index.js:400]: Same.

File: app/javascript/mastodon/features/ui/components/confirmation_modal.js [hunk 5]
```diff
@@ -14,22 +15,41 @@ class ConfirmationModal extends React.PureComponent {
     secondary: PropTypes.string,
     onSecondary: PropTypes.func,
     closeWhenConfirm: PropTypes.bool,
+    destructive: PropTypes.bool,
+
```
Comment (ClearlyClaire) [inline app/javascript/mastodon/features/ui/components/confirmation_modal.js:19]: Why the extra blank line?

File: app/lib/user_settings_decorator.rb [hunk 6]
```diff
@@ -14,32 +14,35 @@ def update(settings)
 
   private
 
+  # rubocop: disable Metrics/AbcSize, Metrics/CyclomaticComplexity, Metrics/PerceivedComplexity
```
Comment (ClearlyClaire) [inline app/lib/user_settings_decorator.rb:17]: I wouldn't enable that. This class definitely needs refactoring. But that's for another PR (e.g. #18368)

**Review by ClearlyClaire — CHANGES_REQUESTED** [review]

File: app/lib/user_settings_decorator.rb (hunk 6 above)
Comment (noellabo) [reply app/lib/user_settings_decorator.rb:17]: A commit that avoids the warning will revert.

File: app/javascript/mastodon/features/ui/components/confirmation_modal.js (hunk 5 above)
Comment (noellabo) [reply app/javascript/mastodon/features/ui/components/confirmation_modal.js:19]: Ow!

**Comment by noellabo** [comment]
- Removed unused code from domain_container
- Passphrases can now be translated
- Conditional judgments that are difficult to understand have been reworded to more general wording
- I removed the code that bypasses the rubocop warnings. These will be resolved in another PR https://github.com/mastodon/mastodon/pull/18368

**Review by ClearlyClaire — DISMISSED** [review]

**Comment by Gargron** [comment]
If the point is to prevent accidental clicks, wouldn't this also work with having a checkbox that says "Yes, block {domain}"?

**Comment by noellabo** [comment]
In the light of real-life cases, that is not enough.

The problem was
- When the user's judgment is declining
- When multiple taps occur randomly

Well, increasing the checkboxes to 3 might be enough, but I think it would be better to be strict enough to require input.

**Comment by connorshea** [comment]
This'd be good to have since I've seen a decent number of people accidentally do this to themselves. A simple checkbox for "Yes, I want to block this whole instance" would work for me, since the main thing is just putting a bit of friction into the process to make sure it's not as easy to shoot yourself in the foot.

**Comment by mjankowski** [comment]
Doing old issue triage -- I'm going to close this since it seems like the consensus of the review was "lets just do a standard confirm dialog"; lack of activity in ~2 years; needs rebase.

That said, discussion was not totally conclusive ... so if you want to nudge forward, please rebase/reopen and in theory someone will review again.

**Pull request closed without merging**